x/data/client/testsuite/cli_test.go

```diff
@@ -0,0 +1,456 @@
+		fmt.Sprintf("--%s=%s", flags.FlagBroadcastMode, flags.BroadcastBlock),
+		fmt.Sprintf("--%s=%s", flags.FlagFees, sdk.NewCoins(sdk.NewCoin(s.cfg.BondDenom, sdk.NewInt(10))).String()),
+	}
+
+	testCases := []struct {
+		name         string
+		args         []string
+		expectErr    bool
+		respType     proto.Message
+		expectedCode uint32
+	}{
+		{
+			"less number of args",
+			append(
+				[]string{val.Address.String()},
+				commonFlags...,
+			),
+			true,
+			nil,
+			0,
+		},
+		{
+			"wrong sender",
+			append(
+				[]string{
+					"wrongSender",
+					cid.String(),
+				},
+				commonFlags...,
+			),
+			true,
+			nil,
+			0,
+		},
+		{
+			"wrong cid",
+			append(
+				[]string{
+					val.Address.String(),
+					"wrongCid",
+				},
+				commonFlags...,
+			),
+			true,
+			nil,
+			0,
+		},
+		{
+			"correct data",
+			append(
+				[]string{
+					val.Address.String(),
+					cid.String(),
+				},
+				commonFlags...,
+			),
+			false,
+			&sdk.TxResponse{},
+			0,
+		},
+		{
+			"duplicate data",
+			append(
+				[]string{
+					val.Address.String(),
+					cid.String(),
+				},
+				commonFlags...,
+			),
+			false,
+			&sdk.TxResponse{},
+			18,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		s.Run(tc.name, func() {
+			cmd := dataclient.MsgAnchorDataCmd()
+
+			out, err := clitestutil.ExecTestCLICmd(clientCtx, cmd, tc.args)
+			if tc.expectErr {
+				s.Require().Error(clientCtx.JSONMarshaler.UnmarshalJSON(out.Bytes(), tc.respType), out.String())
+			} else {
+				s.Require().NoError(err, out.String())
+				s.Require().NoError(clientCtx.JSONMarshaler.UnmarshalJSON(out.Bytes(), tc.respType), out.String())
+
+				txResp := tc.respType.(*sdk.TxResponse)
+				s.Require().Equal(tc.expectedCode, txResp.Code, out.String())
+			}
+		})
+	}
+}
+
+func (s *IntegrationTestSuite) TestGetAnchorDataByCID() {
+	val := s.network.Validators[0]
+	clientCtx := val.ClientCtx
+
+	cid := s.storedCid
+
+	testCases := []struct {
+		name      string
+		malleate  func()
+		args      []string
+		expectErr bool
+		respType  proto.Message
+	}{
+		{
+			"with no args",
+			func() {},
+			[]string{
+				fmt.Sprintf("--%s=json", tmcli.OutputFlag),
+			},
+			true,
+			nil,
+		},
+		{
+			"with non existed cid",
+			func() {
+				cid = s.getCid([]byte("xyzabc"))
+			},
+			[]string{
+				cid.String(),
+				fmt.Sprintf("--%s=json", tmcli.OutputFlag),
+			},
+			true,
+			nil,
+		},
+		{
+			"with correct data",
+			func() {
+				cid = s.storedCid
+			},
+			[]string{
+				cid.String(),
+				fmt.Sprintf("--%s=json", tmcli.OutputFlag),
+			},
+			false,
+			&datatypes.QueryByCidResponse{},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		s.Run(tc.name, func() {
+			tc.malleate()
+			cmd := dataclient.QueryByCidCmd()
+
+			out, err := clitestutil.ExecTestCLICmd(clientCtx, cmd, tc.args)
+			if tc.expectErr {
+				s.Require().Error(clientCtx.JSONMarshaler.UnmarshalJSON(out.Bytes(), tc.respType), out.String())
```

Comment:
imo it's a bit weird to JSON unmarshal an error output. Maybe we could add a `expectErrMsg` field in the test case, and if `expectErr` is true, error should match `expectErrMsg`